docs/grammar/SolidityParser.g4

```diff
@@ -182,7 +182,8 @@ locals[
  freeFunctionDefinition:
  	Function (identifier | Fallback | Receive)
  	LParen (arguments=parameterList)? RParen
- 	stateMutability?
+ 	stateMutability? Suffix?
+ 	| Suffix? stateMutability?
```

Comment:
Does this really do what it's supposed to?
I guess this is meant to read "either ``stateMutability Suffix`` or ``Suffix stateMutability`` while both are optional...
But that's not what I think this says... I think it says:
- An optional stateMutability. Followed by
- Either an optional suffix or an optional suffix. Followed by
- Another optional state mutability.

So this should for example now (incorrectly) accept ``function f() pure suffix pure``...

Reply:
The proper way to do this is probably in the same manner as this is handled in ``contractFunctionDefinition``.

Reply:
> So this should for example now (incorrectly) accept `function f() pure suffix pure`...

Ah, it's missing parentheses.

Reply:
No, wait. I just noticed the extra `?`. It was meant to be:

```
stateMutability Suffix?
| Suffix stateMutability?
```

Looks like my suggestion was not applied correctly in https://github.com/ethereum/solidity/pull/14054#discussion_r1148023318 and our tests did not catch it.

Reply:
> Maybe the nicest way overall to handle this would have been *not* to split contract functions from free functions, but rather just make "being free" a flag like ``elementaryTypeName`` has one...

The reason to split it out was that contract function allow arbitrarily named modifiers and those can be repeated while `suffix` on free functions cannot be repeated.

I guess it's possible to resolve it with a flag, but given how much the modifiers lists on those differ, I thought it was cleaner to separate the definitions.

Reply:
```
stateMutability Suffix?
| Suffix stateMutability?
```

means you *always* have a state mutability - and then either an optional suffix or a non-optional suffix - and then another optional state mutability... also doesn't exactly ring right to me ;-).

Reply:
> And it actually made the grammar more precise. Stuff like virtual or override is not allowed on free function after all.

Which is *not* a parser error... hence, technically, this makes the grammar *less* precise ;-).

Reply:
Grammar should be fixed now.

> Which is not a parser error... hence, technically, this makes the grammar less precise ;-).

You're right. I did not consider that.

BTW: it's not the only thing that is mismatched that way. Just off the top of my head, unary + is not allowed in the grammar. I suspect there's much more of that. We'd need to have tests against it to have any hope of keeping it consistent.

Reply:
That's also why I also did not jump immediately at changing it back :)